File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.swift
import Foundation

var list = [Int]()
while(true)
{
    
    let arr = readLine()?.split(separator: " ").map{Int($0)}
    let num = arr![0]
    if(num == 0){break}
    list = [Int](repeating: 0, count: num!)
    
    
    for i in 1...num!
    {
        list[i-1]=arr![i]!
    }
    print(solve(start: 0, end: num!-1))
}

func solve(start:Int, end:Int)->Int
{
    if(start == end) {return list[start]}
    
    let mid = (start+end)/2
    var result = max(solve(start: start, end: mid) , solve(start: mid+1, end: end))
    
    var left = mid
    var right = mid + 1
    var height = min(list[left], list[right])
    result = max(result, height*2)
    
    while(start < left || right < end)
    {
        if(start < left && right < end)
        {
            if (list[left - 1] < list[right + 1])
            {
                right += 1
                height = min(height, list[right])
                
            }else
            {
                left -= 1
                height = min(height, list[left])
            }
        }
        else if(start < left)
        {
            left -= 1
            height = min(height, list[left])
        }else if (right < end)
        {
            right += 1
            height = min(height, list[right])
        }
        result = max(result, height * (right - left + 1))
    }
    
    return result
}

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
while True:
    n, *heights = map(int, input().split())
    if n == 0:
        break

    stack = []
    max_area = 0

    for i in range(n):
        while stack and heights[stack[-1]] > heights[i]:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)
        stack.append(i)

    while stack:
        height = heights[stack.pop()]
        width = n if not stack else n - stack[-1] - 1
        max_area = max(max_area, height * width)

    print(max_area)
